flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicSink.java

```diff
@@ -83,38 +97,53 @@
 	protected final Properties properties;
 
 	/** Partitioner to select Kafka partition for each item. */
-	protected final Optional<FlinkKafkaPartitioner<RowData>> partitioner;
+	protected final @Nullable FlinkKafkaPartitioner<RowData> partitioner;
 
 	/** Sink commit semantic.*/
 	protected final KafkaSinkSemantic semantic;
 
 	public KafkaDynamicSink(
 			DataType physicalDataType,
+			@Nullable EncodingFormat<SerializationSchema<RowData>> keyEncodingFormat,
+			EncodingFormat<SerializationSchema<RowData>> valueEncodingFormat,
+			int[] keyProjection,
+			int[] valueProjection,
+			@Nullable String keyPrefix,
 			String topic,
 			Properties properties,
-			Optional<FlinkKafkaPartitioner<RowData>> partitioner,
-			EncodingFormat<SerializationSchema<RowData>> encodingFormat,
+			@Nullable FlinkKafkaPartitioner<RowData> partitioner,
 			KafkaSinkSemantic semantic) {
+		// Format attributes
 		this.physicalDataType = Preconditions.checkNotNull(physicalDataType, "Physical data type must not be null.");
+		this.keyEncodingFormat = keyEncodingFormat;
+		this.valueEncodingFormat = Preconditions.checkNotNull(valueEncodingFormat, "Value encoding format must not be null.");
+		this.keyProjection = Preconditions.checkNotNull(keyProjection, "Key projection must not be null.");
+		this.valueProjection = Preconditions.checkNotNull(valueProjection, "Value projection must not be null.");
+		this.keyPrefix = keyPrefix;
+		// Mutable attributes
 		this.metadataKeys = Collections.emptyList();
+		// Kafka-specific attributes
 		this.topic = Preconditions.checkNotNull(topic, "Topic must not be null.");
 		this.properties = Preconditions.checkNotNull(properties, "Properties must not be null.");
-		this.partitioner = Preconditions.checkNotNull(partitioner, "Partitioner must not be null.");
-		this.encodingFormat = Preconditions.checkNotNull(encodingFormat, "Encoding format must not be null.");
+		this.partitioner = partitioner;
 		this.semantic = Preconditions.checkNotNull(semantic, "Semantic must not be null.");
 	}
 
 	@Override
 	public ChangelogMode getChangelogMode(ChangelogMode requestedMode) {
-		return this.encodingFormat.getChangelogMode();
+		return valueEncodingFormat.getChangelogMode();
 	}
 
 	@Override
 	public SinkRuntimeProvider getSinkRuntimeProvider(Context context) {
+		final SerializationSchema<RowData> keySerialization =
+				createSerialization(context, keyEncodingFormat, keyProjection, keyPrefix);
+
 		final SerializationSchema<RowData> valueSerialization =
-				this.encodingFormat.createRuntimeEncoder(context, this.physicalDataType);
+				createSerialization(context, valueEncodingFormat, valueProjection, keyPrefix);
```

Comment:
Good review! This was one of the things where I was struggling. In the end, I decided for removing the prefix also from the values. IMHO the `keyPrefix` is just an intermediate helper option that should not affect the serialization data type. I think no user wants to have `user_id, key_name, payload` in the values. Esp. when `value.fields-include = ALL` we need to strip the prefix also from the value, otherwise a "logical" concept ends up in the physical data.

Reply:
What do you think about:

when `'key.fields-prefix'` is defined, `value.fields-include` shoudn't be defined to `ALL`, and all value fields shoudn't be prefixed with  `keyPrefix`. 

I think `'key.fields-prefix'` is only used when users want read out key and value parts separately and avoid name clashes.

Reply:
Yes, I'm fine with this limitation. I will update the PR.